### PR TITLE
ENH: Use pyarrow.compute for unique, dropna

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
Thanks @mroeschke for the PR. my understanding of #42613 is that we should no longer be implementing any fallback behavior. (there is not definitive policy on that there, so have not yet removed the fallbacks already implemented for StringArray)

That issue applies specifically to StringArray but with the work you have done/doing to have a common base class for pyarrow backed EAs, we may be adding fallback behaviour for the pyarrow backed StringArray?

**Comment by simonjayhawkins** [comment]
> my understanding of #42613 is that ...

more of the discussion was actually in #42597

**Comment by mroeschke** [comment]
Ah thanks, I wasn't aware of this discussion.

My prior, related PR's so far have just been moving existing methods, so no fallback behavior should have been introduced I think.

For this PR, I will just remove the super calls for now and reintroduce them later if we decide on a different fallback policy

**Comment by jreback** [comment]
> Ah thanks, I wasn't aware of this discussion.
> 
> My prior, related PR's so far have just been moving existing methods, so no fallback behavior should have been introduced I think.
> 
> For this PR, I will just remove the super calls for now and reintroduce them later if we decide on a different fallback policy

comments on the other issues is we can show a PerformanceWarning if we are falling back. But let's do that as a pre-cursor

File: pandas/tests/base/test_unique.py [hunk 1]
```diff
@@ -9,10 +14,20 @@
 from pandas.tests.base.common import allow_na_ops
 
 
+def maybe_perf_warn(using_pyarrow):
```
Comment (jreback) [inline pandas/tests/base/test_unique.py:17]: ideally move to the _test_decorators.py (or similar) e.g. this is a general testing function.

**Review by jreback — APPROVED** [review]
lgtm. if you can move the warning tester to a more general location in a followup

**Comment by jreback** [comment]
actually if you can do that move in this PR and I think this also needs a whatsnew note. ping on green.

File: pandas/tests/base/test_unique.py [hunk 2]
```diff
@@ -12,7 +15,11 @@
 def test_unique(index_or_series_obj):
     obj = index_or_series_obj
     obj = np.repeat(obj, range(1, len(obj) + 1))
-    result = obj.unique()
+    with tm.maybe_produces_warning(
+        PerformanceWarning,
+        pa_version_under2p0 and str(index_or_series_obj.dtype) == "string[pyarrow]",
```
Comment (simonjayhawkins) [inline pandas/tests/base/test_unique.py:20]: It appears that with a pyarrow backed StringSrray, we are only testing Index here, not Series?

Also don't need the `str` cast, dtype equality to the string form should work.

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @mroeschke lgtm except some questions. any benchmark results?

File: pandas/tests/base/test_unique.py (hunk 2 above)
Comment (mroeschke) [reply pandas/tests/base/test_unique.py:20]: > It appears that with a pyarrow backed StringSrray, we are only testing Index here, not Series?

Looks so based on the fixture in `pandas/conftest.py`

```
if has_pyarrow:
    idx = Index(pd.array(tm.makeStringIndex(100), dtype="string[pyarrow]"))
    indices_dict["string-pyarrow"] = idx
```

Fixed the comparison

File: pandas/core/arrays/arrow/array.py [hunk 3]
```diff
@@ -104,6 +107,20 @@ def copy(self: ArrowExtensionArrayT) -> ArrowExtensionArrayT:
...
+
+        Returns
+        -------
+        valid : ArrowExtensionArray
+        """
+        if pa_version_under6p0:
+            fallback_performancewarning(version="6")
+            return super().dropna()
+        else:
+            return type(self)(pc.drop_null(self._data))
```
[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/core/arrays/arrow/array.py:122]: Hmm not exactly sure what you mean here.

File: pandas/tests/indexes/test_common.py [hunk 4]
```diff
@@ -395,7 +398,9 @@ def test_astype_preserves_name(self, index, dtype):
 
         try:
             # Some of these conversions cannot succeed so we use a try / except
-            with tm.assert_produces_warning(warn):
+            with tm.assert_produces_warning(
+                warn, raise_on_extra_warnings=not pa_version_under2p0
```
[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/tests/indexes/test_common.py:402]: Fixed

**Comment by mroeschke** [comment]
Here are the perf differences (I think I see what you mean about the dropna dispatch)

PR
```
In [1]: import pyarrow as pa
   ...: pa.__version__
   ...:
Out[1]: '7.0.0'

In [2]: ser = pd.Series(["1", np.nan] * 100_000, dtype="string[pyarrow]")

In [4]: %timeit ser.unique()
   ...:
1.3 ms ± 6.59 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [5]: %timeit ser.dropna()

2.27 ms ± 46.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

main
```
In [1]: ser = pd.Series(["1", np.nan] * 100_000, dtype="string[pyarrow]")
   ...:

In [2]: %timeit ser.unique()
   ...:
24.4 ms ± 181 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [3]: %timeit ser.dropna()
2.27 ms ± 14.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

File: pandas/core/arrays/arrow/array.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/core/arrays/arrow/array.py:122]: Ah I see now. Yeah hooking this up to dropna might be a good idea in a future PR

**Review by simonjayhawkins — APPROVED** [review]
Thanks @mroeschke lgtm

on the casting to string for the dtype equality in the tests, I thought that we should be able to create a pyarrow StringDtype (but not an ArrowStringArray) without pyarrow installed, but I must be wrong here.

Also I think the index_or_series_obj fixture should include arrow backed Series, we don't have any testing for this at the moment. (or for DataFrame)

**Comment by jreback** [comment]
thanks @mroeschke really nice

**Pull request merged**